### PR TITLE
overlord/ifacemgr: basic measurements

### PR DESCRIPTION
**Pull request by stolowski**
Add two helpers related to Timings (`NewForTask` and `Run`) and a few points for collecting Timings - for starters these are a few task handlers and helpers of ifacemgr, startup code of ifacemgr, ~~and Ensure loop of snapmgr~~. These will currently generate too many irrelevant timings - one idea is to not Save a Timings tree if it's total execution is less than a given _duration threshold_ (with an enhanced version of `Save`).

This is mostly RFC, we need to decide on the wording, labels/summaries etc, what to record and what to ignore etc.

These timings will be available via a (new?) variant of `snap debug timings` command. Sample output is here - https://paste.ubuntu.com/p/gXrk8sBwwB/ (this requires client and api bits which I've in a separate branch).

### CODE REVIEW

File: overlord/snapstate/snapmgr.go [hunk 1]
```diff
@@ -644,26 +645,37 @@ func (m *SnapManager) localInstallCleanup() error {
...
+		err := f()
+		meas.Stop()
+		return err
+	}
+
 	// do not exit right away on error
 	errs := []error{
 		m.atSeed(),
-		m.ensureAliasesV2(),
+		measure(m.ensureAliasesV2, "aliases", ""),
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go]: I don't  think measuring at this granularity is that useful, also it creates the problem that we get very short times most of the time, I would  undo the changes here for this PR and we can discuss what to do about Ensures separately.

Comment (stolowski) [reply overlord/snapstate/snapmgr.go]: Done.

**Comment by stolowski** [comment]
@pedronis I made a rough estimate of the size of the serialized json data for the timings proposed in this PR (no snapmgr ensure, just ifacemgr timings): new snapd install, the following snaps get installed:
```
$ snap list
Name        Version    Rev   Tracking  Publisher     Notes
core        16-2.33.1  4917  stable    canonical✓    core
core18      18         731   stable    canonical✓    base
htop        2.2.0      1168  stable    maxiberta     -
obs-studio  23.0.2     345   stable    snapcrafters  -
```

The complete timings tree printed by the client: 
https://paste.ubuntu.com/p/CQBVT6Qzvd/

This is 20 Timings, with a few nested Spans each. Their total size (serialized json) is ~46kB:
```
$ sudo cat /var/lib/snapd/state.json | jq -r --compact-output '.data["timings"]' | wc -c
47297
```

File: timings/helpers.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	tags := map[string]string{"task-id": task.ID()}
+	if chg := task.Change(); chg != nil {
+		tags["change-id"] = chg.ID()
+	}
+	return New(tags)
+}
+
+// Run creates, starts and then stops a nested Span under parent span. The nested
+// Span is passed to the measured function and it can used to create further spans.
+func (t *Span) Run(label, summary string, f func(nestedTiming Measurer)) {
```
Comment (pedronis) [inline timings/helpers.go]: I would slightly prefer  this to be one function:

timing.Run(meas Measurer,  label, summary string, f func(nestedTiming Measurer))

it's slightly longer but is easier to grep, under what is about. There are many Run.

File: overlord/ifacestate/handlers.go [hunk 3]
```diff
@@ -78,6 +79,9 @@ func (m *InterfaceManager) doSetupProfiles(task *state.Task, tomb *tomb.Tomb) er
 	task.State().Lock()
 	defer task.State().Unlock()
 
+	perfTimings := timings.NewForTask(task)
+	defer perfTimings.Save(task.State()) // XXX: should the error from save be returned from handler?
```
Comment (pedronis) [inline overlord/ifacestate/handlers.go]: maybe Save shouldn't return an error, but log direclty and that's it?  we would likely not return the error anyway here but log it

Comment (stolowski) [reply overlord/ifacestate/handlers.go]: Yes, I think that's the right approach, thanks, done.

**Comment by stolowski** [comment]
Some timings from rpi3: https://pastebin.ubuntu.com/p/5TDqz2s7wj/

**Comment by stolowski** [comment]
After the simplifications discussed so far and applying duration threshold the timings similiar to the earlier scenario take approximately 13kB, and the output looks as follows: https://paste.ubuntu.com/p/rtr4QFVtXN/

File: overlord/ifacestate/handlers.go [hunk 4]
```diff
@@ -195,25 +199,29 @@ func (m *InterfaceManager) setupProfilesForSnap(task *state.Task, _ *tomb.Tomb,
...
+
+	return m.setupSecurityByBackend(task, affectedSnaps, confinementOpts, tm)
 }
 
 func (m *InterfaceManager) doRemoveProfiles(task *state.Task, tomb *tomb.Tomb) error {
 	st := task.State()
 	st.Lock()
 	defer st.Unlock()
 
+	perfTimings := timings.NewForTask(task)
```
Comment (pedronis) [inline overlord/ifacestate/handlers.go]: I didn't think of this so far, but should we capture task status to be able to distinguish doing from undoing?

File: overlord/ifacestate/ifacemgr.go [hunk 5]
```diff
@@ -53,6 +54,8 @@ type InterfaceManager struct {
 // Manager returns a new InterfaceManager.
 // Extra interfaces can be provided for testing.
 func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.TaskRunner, extraInterfaces []interfaces.Interface, extraBackends []interfaces.SecurityBackend) (*InterfaceManager, error) {
+	perftimings := timings.New(map[string]string{"startup": "interface manager"})
```
Comment (pedronis) [inline overlord/ifacestate/ifacemgr.go]: this should go after after the delayedCrossMgrInit

"startup": "ifacmgr"  ?  

s/perftimings/perfTimings/ ?

File: interfaces/apparmor/backend.go [hunk 6]
```diff
@@ -395,15 +396,24 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
 	for i, profile := range changed {
 		pathnames[i] = filepath.Join(dir, profile)
 	}
-	errReloadChanged := loadProfiles(pathnames, cache, skipReadCache)
+
+	var errReloadChanged error
+	timings.Run(tm, "load-profiles", fmt.Sprintf("load changed security profiles of snap %q", snapInfo.InstanceName()), func(nesttm timings.Measurer) {
+		errReloadChanged = loadProfiles(pathnames, cache, skipReadCache)
```
Comment (pedronis) [inline interfaces/apparmor/backend.go]: I wonder if we should distinguish this one from the one below?

"load-profiles[changed]"  vs "load-profiles[unchanged]"  perhaps ?

File: overlord/ifacestate/handlers.go (hunk 4 above)
Comment (stolowski) [reply overlord/ifacestate/handlers.go]: Done.

File: interfaces/apparmor/backend.go (hunk 6 above)
Comment (stolowski) [reply interfaces/apparmor/backend.go]: Good idea, done.

File: overlord/ifacestate/handlers.go [hunk 7]
```diff
@@ -78,6 +79,9 @@ func (m *InterfaceManager) doSetupProfiles(task *state.Task, tomb *tomb.Tomb) er
 	task.State().Lock()
 	defer task.State().Unlock()
 
+	perfTimings := timings.NewForTask(task)
+	defer perfTimings.Save(task.State())
```
Comment (mvo5) [inline overlord/ifacestate/handlers.go]: I like the API

File: overlord/ifacestate/helpers.go [hunk 8]
```diff
@@ -197,12 +198,14 @@ func (m *InterfaceManager) regenerateAllSecurityProfiles() error {
 				continue // Test backends have no name, skip them to simplify testing.
 			}
 			// Refresh security of this snap and backend
-			if err := backend.Setup(snapInfo, opts, m.repo); err != nil {
-				// Let's log this but carry on without writing the system key.
-				logger.Noticef("cannot regenerate %s profile for snap %q: %s",
-					backend.Name(), snapName, err)
-				shouldWriteSystemKey = false
-			}
+			timings.Run(tm, "setup-security-backend", fmt.Sprintf("setup security backend %q for snap %q", backend.Name(), snapInfo.InstanceName()), func(nesttm timings.Measurer) {
```
Comment (mvo5) [inline overlord/ifacestate/helpers.go]: Can we test in ifacemgr_test.go (or somewhere) that the timings are actually run? Or does that not make much sense?

File: interfaces/apparmor/backend_test.go [hunk 9]
```diff
@@ -179,7 +184,7 @@ func (s *backendSuite) TestProfilesAreAlwaysLoaded(c *C) {
 	for _, opts := range testedConfinementOpts {
 		snapInfo := s.InstallSnap(c, opts, "", ifacetest.SambaYamlV1, 1)
 		s.parserCmd.ForgetCalls()
-		err := s.Backend.Setup(snapInfo, opts, s.Repo)
+		err := s.Backend.Setup(snapInfo, opts, s.Repo, s.meas)
```
[the remark this answers is not in the document]
Comment (stolowski) [reply interfaces/apparmor/backend_test.go]: That's probably a good idea, added. It's not testing every aspect of the serialized timings as these are alredy tested in detail in timings package.

File: overlord/ifacestate/helpers.go (hunk 8 above)
Comment (stolowski) [reply overlord/ifacestate/helpers.go]: Added.
